src/merkle_tree/mod.rs

```diff
@@ -181,6 +183,212 @@ impl<P: Config> Path<P> {
+        for path in auth_paths[1..].to_vec() {
+            let mut prefix_len = 0;
+            if prev_path.auth_path.len() != 0 && path.auth_path.len() != 0 {
+                while prev_path.auth_path[prefix_len] == path.auth_path[prefix_len] {
+                    prefix_len += 1;
+                    if prefix_len == prev_path.auth_path.len() || prefix_len == path.auth_path.len()
+                    {
+                        break;
+                    }
+                }
+            }
+            auth_paths_prefix_lenghts.push(prefix_len);
+            if prefix_len == prev_path.auth_path.len() {
+                auth_paths_suffixes.push(vec![]);
+            } else {
+                auth_paths_suffixes.push(path.auth_path[prefix_len..].to_vec().clone());
+            }
+            prev_path = path;
+        }
+
+        Ok(MultiPath {
+            leaf_indexes: indexes,
+            auth_paths_prefix_lenghts,
+            auth_paths_suffixes,
+            leaf_siblings_hashes,
+        })
+    }
+
+    /// Returns the decompressed authentication paths for every leaf index
+    fn decompress(
+        &'_ self,
+    ) -> Result<impl '_ + Iterator<Item = Vec<P::InnerDigest>>, crate::Error> {
+        // Incrementally reconstruct all the paths
+        let mut curr_path = self.auth_paths_suffixes[0].clone();
+        let mut auth_paths = (0..self.leaf_indexes.len())
+            .map(|_| Vec::with_capacity(curr_path.len()))
+            .collect::<Vec<Vec<P::InnerDigest>>>();
+        auth_paths[0] = self.auth_paths_suffixes[0].clone();
+
+        for i in 1..auth_paths.len() {
+            auth_paths[i].extend_from_slice(&curr_path[0..self.auth_paths_prefix_lenghts[i]]);
+            auth_paths[i].extend(self.auth_paths_suffixes[i].clone());
+            curr_path = auth_paths[i].clone();
+        }
+        Ok(auth_paths.into_iter())
+    }
+
+    /// Verify that leaves are at `self.leaf_indexes` of the merkle tree.
+    /// Note that the order of the leaves hashes should match the leaves respective indexes
+    /// * `leaf_size`: leaf size in number of bytes
+    ///
+    /// `verify` infers the tree height by setting `tree_height = self.auth_paths_suffixes[0].len() + 2`
+    pub fn verify<L: Borrow<P::Leaf> + Clone>(
+        &self,
+        leaf_hash_params: &LeafParam<P>,
+        two_to_one_params: &TwoToOneParam<P>,
+        root_hash: &P::InnerDigest,
+        leaves: impl IntoIterator<Item = L>,
+    ) -> Result<bool, crate::Error> {
+        // array of auth paths as arrays of InnerDigests
+
+        let mut leaves = leaves.into_iter();
+
+        let mut auth_paths = self.decompress()?.peekable();
+
+        let tree_height = auth_paths.peek().unwrap().len() + 2;
+
+        // LookUp table to speedup computation avoid redundant hash computations
+        let mut hash_lut: HashMap<usize, P::InnerDigest> = HashMap::new();
+
+        for i in 0..self.leaf_indexes.len() {
```

Comment:
This I wonder if there's any way to avoid calling the decompression step explicitly?

It looks like the loop that's called inside `decompress` is also going over all the leaf indices, just like you do in this loop inside `verify`.